app/src/main/java/com/eveningoutpost/dexdrip/cgm/connectfollow/ConnectFollowDownloader.java

```diff
@@ -0,0 +1,207 @@
+package com.eveningoutpost.dexdrip.cgm.connectfollow;
+
+import android.os.PowerManager;
+
+import com.eveningoutpost.dexdrip.Models.JoH;
+import com.eveningoutpost.dexdrip.Models.UserError;
+import com.eveningoutpost.dexdrip.R;
+import com.eveningoutpost.dexdrip.UtilityModels.CollectionServiceStarter;
+import com.eveningoutpost.dexdrip.UtilityModels.Constants;
+import com.eveningoutpost.dexdrip.UtilityModels.Inevitable;
+import com.eveningoutpost.dexdrip.cgm.connectfollow.messages.ConnectData;
+import com.eveningoutpost.dexdrip.cgm.connectfollow.messages.ConnectDataResult;
+import com.eveningoutpost.dexdrip.xdrip;
+
+import java.util.List;
+
+import lombok.Getter;
+
+import static com.eveningoutpost.dexdrip.Models.JoH.emptyString;
+
+/**
+ * Medtronic CareLink Connect follow
+ */
+
+
+public class ConnectFollowDownloader {
+
+    private static final String TAG = "ConnectFollowDL";
+    private static final boolean D = false;
+
+    private String carelinkUsername;
+    private String carelinkPassword;
+    private Country carelinkCountry;
+
+    private ConnectClient connectClient;
+
+    private boolean loginDataLooksOkay;
+
+    private static PowerManager.WakeLock wl;
+
+    private String status;
+
+    private long loginBlockedTill = 0;
+    private long loginBackoff = Constants.MINUTE_IN_MS;
+
+    public String getStatus(){
+        return status;
+    }
+
+    ConnectFollowDownloader(String carelinkUsername, String carelinkPassword, Country carelinkCountry) {
+        this.carelinkUsername = carelinkUsername;
+        this.carelinkPassword = carelinkPassword;
+        this.carelinkCountry = carelinkCountry;
+        loginDataLooksOkay = !emptyString(carelinkUsername) && !emptyString(carelinkPassword) && carelinkCountry != null && carelinkCountry != Country.Invalid;
+    }
+
+    public static void resetInstance() {
+        //retrofit = null;
+        //service = null;
+        UserError.Log.d(TAG, "Instance reset");
+        CollectionServiceStarter.restartCollectionServiceBackground();
+    }
+
+    public boolean doEverything( ) {
+        msg("Start download");
+
+        if (D) UserError.Log.e(TAG, "doEverything called");
+        if (loginDataLooksOkay) {
+            //TODO
```

Comment:
What has to be done, here?

Reply:
Differentiate between various communication error types and use different strategies  on how to continue, how much time to wait for next, when to reset temporary wait time etc. The communication works now as well, but maybe it could be reacting better to communication issues.

Reply:
I just meant the "TODO". Every todo in the code means it is incomplete.